### ISSUE TITLE
M3 Ultra 512GB benchmark results: training 110M–5B, forward 5B–25B

### ISSUE THREAD

**Issue by A-Beata** [ncdrone/rustane#16]
## Hardware

- **Chip**: Apple M3 Ultra (2 ANE dies)
- **RAM**: 512 GB
- **macOS**: 26.3.1 (Build 25D2128)
- **Model**: Mac Studio (Mac15,14)
- **Branch**: `m3-ultra` @ commit f0fa1f7 (includes decomposed FFN fix from PR #15)

## Training Pipeline Validation (110M–1.5B)

All configs: compile → forward → 10-step training (loss must decrease) → timing.

```
=== 110M — 1024d/2816h/8L/seq512 — ~111M params ===
  [1/4] Compiling 10 ANE kernels... OK (3.0s)
  [2/4] Forward pass... OK (loss=9.0170)
  [3/4] Training 10 steps... 9.0170 → 8.9893 (delta=-0.0277)
  [4/4] Timing 3 steps... 293ms/step (median of 3)
  PASS: 110M

=== 600M — 1536d/4096h/20L/seq512 — ~579M params ===
  [1/4] Compiling 10 ANE kernels... OK (3.0s)
  [2/4] Forward pass... OK (loss=9.0108)
  [3/4] Training 10 steps... 9.0108 → 8.9857 (delta=-0.0251)
  [4/4] Timing 3 steps... 1302ms/step (median of 3)
  PASS: 600M

=== 800M — 1792d/4864h/24L/seq512 — ~951M params ===
  [1/4] Compiling 10 ANE kernels... OK (3.2s)
  [2/4] Forward pass... OK (loss=9.0178)
  [3/4] Training 10 steps... 9.0178 → 8.9960 (delta=-0.0218)
  [4/4] Timing 3 steps... 1794ms/step (median of 3)
  PASS: 800M

=== 1B — 2048d/5632h/28L/seq512 — ~1456M params ===
  [1/4] Compiling 10 ANE kernels... OK (3.2s)
  [2/4] Forward pass... OK (loss=9.0178)
  [3/4] Training 10 steps... 9.0178 → 8.9958 (delta=-0.0220)
  [4/4] Timing 3 steps... 2914ms/step (median of 3)
  PASS: 1B

=== 1.5B — 2304d/6144h/32L/seq512 — ~2057M params ===
  [1/4] Compiling 10 ANE kernels... OK (3.0s)
  [2/4] Forward pass... OK (loss=9.0257)
  [3/4] Training 10 steps... 9.0257 → 9.0023 (delta=-0.0234)
  [4/4] Timing 3 steps... 3844ms/step (median of 3)
  PASS: 1.5B
```

**Note:** 1B and 1.5B use the decomposed FFN path (PR #15) since `2*seq + 3*hidden > 16,384`. Below 1B uses the fused path unchanged.

## Stress Test: Full Training Step (2B–5B)

Full training step: forward + backward + Metal Adam optimizer.

```
--- 2B: 2.56B params, 2560d/6912h/32L ---
    [1] Compile... OK (1.0s)
    [2] Allocate... OK (1.6s)
    [3] Forward... OK (loss=9.0354, 1884ms)
    [4] Backward... OK (3069ms)
    [5] Full step... OK (5558ms/step = 1884 fwd + 3069 bwd + 605 upd)
    PASS: 2B at 2.56B — 5558ms/step

--- 3B: 3.50B params, 2816d/7680h/36L ---
    [1] Compile... OK (0.9s)
    [2] Allocate... OK (2.2s)
    [3] Forward... OK (loss=9.0447, 2450ms)
    [4] Backward... OK (4120ms)
    [5] Full step... OK (7544ms/step = 2450 fwd + 4120 bwd + 974 upd)
    PASS: 3B at 3.50B — 7544ms/step

--- 4B: 4.10B params, 3072d/8192h/36L ---
    [1] Compile... OK (1.0s)
    [2] Allocate... OK (2.6s)
    [3] Forward... OK (loss=9.0652, 3023ms)
    [4] Backward... OK (5844ms)
    [5] Full step... OK (9828ms/step = 3023 fwd + 5844 bwd + 961 upd)
    PASS: 4B at 4.10B — 9828ms/step

--- 5B: 5.38B params, 3328d/8960h/40L ---
    [1] Compile... OK (1.5s)
    [2] Allocate... OK (3.4s)
    [3] Forward... OK (loss=9.0959, 3885ms)
    [4] Backward... OK (6689ms)
    [5] Full step... OK (11885ms/step = 3885 fwd + 6689 bwd + 1311 upd)
    PASS: 5B at 5.38B — 11885ms/step

=== ALL PASSED through 5B! ===
```

## Training at 10B

10B compiles and runs forward+backward successfully but loss does not decrease over 10 steps:

```
=== 10B — 4096d/11008h/48L/seq512 — ~9748M params ===
  [1/4] Compiling ANE kernels... 1.9s
  [2/4] Forward pass... loss=9.3228
  [3/4] Training 10 steps... 9.3228 → 9.4113 (delta=+0.0885)
  [4/4] Timing 5 steps... 19925ms/step (fwd=5784 bwd=12945 upd=870) = 26 tok/s
```

Kernels compile and execute, but training convergence fails. Likely needs hyperparameter tuning at this scale (learning rate, warmup, grad clip).

## Forward-Only Scale Ladder (5B–20B)

```
====================================================================================================
  FORWARD-ONLY SCALE RESULTS
====================================================================================================
  name           params    dim    hid   nl  compile    alloc        fwd     loss  est.RAM
  -----------------------------------------------------------------------------------------------
  5B              5.01B   3072   8192   44     1.1s     3.2s      3214ms   9.0652    25.2GB
  7B              6.51B   4096  11008   32     1.8s     4.1s      3817ms   9.3228    31.1GB
  10B             9.75B   4096  11008   48     1.9s     6.0s      5750ms   9.3228    46.4GB
  13B            12.73B   5120  13824   40     2.9s     8.0s      8999ms  10.2874    58.6GB
  15B            15.27B   5120  13824   48     3.0s     9.5s     10784ms  10.2874    70.2GB
  20B            20.34B   5120  13824   64     2.9s    12.7s     14374ms  10.2874    93.3GB
```

## Forward-Only Ceiling (25B–30B)

```
======================================================================
  25B — 5120d/13824h/80L/seq512 — 25.42B params — est. 116.5GB
======================================================================
  [1/3] Compiling ANE kernels... 3.0s
  [2/3] Allocating 116.5GB... 16.0s
  [3/3] Forward pass (1 warmup + 3 timed)... 18007ms (loss=10.2874)

======================================================================
  30B — 6144d/16384h/64L/seq512 — 29.04B params — est. 130.4GB
======================================================================
  [1/3] Compiling ANE kernels... FAILED
  sdpaFwd compile: _ANECompiler : ANECCompile() FAILED
```

**25B passes. 30B fails at sdpaFwd compilation (dim=6144 hits ANE compiler limit).**

### Training (full step)

| Scale | M3 Ultra 512GB | M4 Max 128GB |
|-------|---------------|-------------|
| 110M | 293ms | — |
| 600M | 1,302ms | — |
| 800M | 1,794ms | — |
| 1B | 2,914ms | — |
| 1.5B | 3,844ms | — |
| 2B | 5,558ms | — |
| 3B | 7,544ms | — |
| 4B | 9,828ms | — |
| 5B | 11,885ms | — |
| 10B | 19,925ms (no convergence) | OOM |

### Forward-only

| Scale | M3 Ultra 512GB | M4 Max 128GB | Speedup |
|-------|---------------|-------------|---------|
| 5B | 3,214ms | 2,064ms | 0.64x (slower) |
| 7B | 3,817ms | 3,132ms | 0.82x |
| 10B | 5,750ms | 4,696ms | 0.82x |
| 13B | 8,999ms | 21,962ms | **2.44x** |
| 15B | 10,784ms | 26,740ms | **2.48x** |
| 20B | 14,374ms | 40,933ms | **2.85x** |
| 25B | 18,007ms | — (OOM) | — |

## Key Findings

1. **Dual-die advantage only kicks in at dim=5120+**: Below 5120 (5B–10B), Ultra is actually slower than Max (no dual-die scheduling observed). At dim=5120+ (13B+), Ultra is 2.4–2.8x faster — suggesting Apple's ANE scheduler splits work across dies when matrices are large enough.

2. **dim=5120 efficiency cliff is absent on Ultra**: M4 Max had a 4.7x/layer cliff at dim=5120. The Ultra's dual-die scheduling appears to eliminate this bottleneck entirely.

3. **ANE compilation ceiling at dim=6144**: 30B (dim=6144) fails at sdpaFwd compilation, not ffnFused. This is a separate ANE compiler limit from the spatial width issue fixed in PR #15.

4. **Training works through 5B, 10B kernels compile but need tuning**: Forward+backward execute correctly at 10B. Loss not decreasing is likely a hyperparameter issue, not a hardware/kernel issue.

5. **Forward ceiling: 25B on 512GB**: 25.4B params forward pass completes in 18s. RAM is no longer the bottleneck on 512GB — ANE compiler limits are.
